projects/button_challenge/src/main.c

```diff
@@ -0,0 +1,59 @@
+#include <stdint.h>
+#include "adc.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "interrupt.h"
+#include "log.h"
+
+// Addresses
+static const GpioAddress button_address = { .port = GPIO_PORT_B, .pin = 2 };
+
+static const GpioAddress adc_address = { .port = GPIO_PORT_A, .pin = 6 };
+
+// Settings
+static GpioSettings button_settings = { .direction = GPIO_DIR_IN,
+                                        .state = GPIO_STATE_HIGH,
+                                        .alt_function = GPIO_ALTFN_NONE,
+                                        .resistor = GPIO_RES_PULLDOWN };
+
+static GpioSettings adc_settings = { .direction = GPIO_DIR_OUT,
+                                     .state = GPIO_STATE_LOW,
+                                     .alt_function = GPIO_ALTFN_ANALOG,
+                                     .resistor = GPIO_RES_NONE };
+
+static InterruptSettings interrupt_settings = { .type = INTERRUPT_TYPE_INTERRUPT,
+                                                .priority = INTERRUPT_PRIORITY_NORMAL };
+
+// Callback function
+static void prv_adc_callback(const GpioAddress *address, void *context) {
+  AdcChannel *adc_channel = context;
+  uint16_t adc_data = 0;
+
+  adc_read_raw(*adc_channel, &adc_data);
+  LOG_DEBUG("ADC data: %d\n", adc_data);
+}
+
+// Main
+int main() {
+  interrupt_init();
+  gpio_init();
+  gpio_it_init();
+
+  gpio_init_pin(&button_address, &button_settings);
+  gpio_init_pin(&adc_address, &adc_settings);
+  adc_init(ADC_MODE_SINGLE);
+
+  AdcChannel adc_channel = NUM_ADC_CHANNELS;
+
+  adc_get_channel(adc_address, &adc_channel);
```

Comment:
Using channels is outdated, though this looks okay.
I would recommend looking at `adc.h` again. There is a better way to do this so you don't have to manually set the channels.